File: bindings/rust/src/netconf.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! NETCONF types for the proven-servers ABI.
//!
//! Formally verified NETCONF types (RFC 6241).
//! Mirrors the Idris2 module `NetconfABI.Types`.
//!
//! - `NetconfOperation` -- NETCONF operations.
//! - `Datastore` -- NETCONF datastores.
//! - `EditOperation` -- NETCONF edit operations.
//! - `NetconfErrorType` -- NETCONF error types.
//! - `ErrorSeverity` -- NETCONF error severity.
//! - `NetconfState` -- NETCONF session states.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// NETCONF Constants
// ===========================================================================

/// Standard NETCONF SSH port.
pub const NETCONF_PORT: u16 = 830;

// ===========================================================================
// NetconfOperation (tags 0-11)
// ===========================================================================

/// NETCONF operations.
///
/// Matches `NetconfOperation` in `NetconfABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum NetconfOperation {
    /// Get (tag 0).
    Get = 0,
    /// GetConfig (tag 1).
    GetConfig = 1,
    /// EditConfig (tag 2).
    EditConfig = 2,
    /// CopyConfig (tag 3).
    CopyConfig = 3,
    /// DeleteConfig (tag 4).
    DeleteConfig = 4,
    /// Lock (tag 5).
    Lock = 5,
    /// Unlock (tag 6).
    Unlock = 6,
    /// CloseSession (tag 7).
    CloseSession = 7,
    /// KillSession (tag 8).
    KillSession = 8,
    /// Commit (tag 9).
    Commit = 9,
    /// Validate (tag 10).
    Validate = 10,
    /// DiscardChanges (tag 11).
    DiscardChanges = 11,
}

impl NetconfOperation {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Get),
            1 => Some(Self::GetConfig),
            2 => Some(Self::EditConfig),
            3 => Some(Self::CopyConfig),
            4 => Some(Self::DeleteConfig),
            5 => Some(Self::Lock),
            6 => Some(Self::Unlock),
            7 => Some(Self::CloseSession),
            8 => Some(Self::KillSession),
            9 => Some(Self::Commit),
            10 => Some(Self::Validate),
            11 => Some(Self::DiscardChanges),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [NetconfOperation; 12] = [
        Self::Get, Self::GetConfig, Self::EditConfig, Self::CopyConfig, Self::DeleteConfig, Self::Lock, Self::Unlock, Self::CloseSession, Self::KillSession, Self::Commit, Self::Validate, Self::DiscardChanges,
    ];
}

impl fmt::Display for NetconfOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Datastore (tags 0-2)
// ===========================================================================

/// NETCONF datastores.
///
/// Matches `Datastore` in `NetconfABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Datastore {
    /// Running (tag 0).
    Running = 0,
    /// Startup (tag 1).
    Startup = 1,
    /// Candidate (tag 2).
    Candidate = 2,
}

impl Datastore {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Running),
            1 => Some(Self::Startup),
            2 => Some(Self::Candidate),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [Datastore; 3] = [
        Self::Running, Self::Startup, Self::Candidate,
    ];
}

impl fmt::Display for Datastore {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// EditOperation (tags 0-4)
// ===========================================================================

/// NETCONF edit operations.
///
/// Matches `EditOperation` in `NetconfABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum EditOperation {
    /// Merge (tag 0).
    Merge = 0,
    /// Replace (tag 1).
    Replace = 1,
    /// Create (tag 2).
    Create = 2,
    /// Delete (tag 3).
    Delete = 3,
    /// Remove (tag 4).
    Remove = 4,
}

impl EditOperation {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Merge),
            1 => Some(Self::Replace),
            2 => Some(Self::Create),
            3 => Some(Self::Delete),
            4 => Some(Self::Remove),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [EditOperation; 5] = [
        Self::Merge, Self::Replace, Self::Create, Self::Delete, Self::Remove,
    ];
}

impl fmt::Display for EditOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// NetconfErrorType (tags 0-3)
// ===========================================================================

/// NETCONF error types.
///
/// Matches `NetconfErrorType` in `NetconfABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum NetconfErrorType {
    /// Transport (tag 0).
    Transport = 0,
    /// RPC (tag 1).
    Rpc = 1,
    /// Protocol (tag 2).
    Protocol = 2,
    /// Application (tag 3).
    Application = 3,
}

impl NetconfErrorType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Transport),
            1 => Some(Self::Rpc),
            2 => Some(Self::Protocol),
            3 => Some(Self::Application),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [NetconfErrorType; 4] = [
        Self::Transport, Self::Rpc, Self::Protocol, Self::Application,
    ];
}

impl fmt::Display for NetconfErrorType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ErrorSeverity (tags 0-1)
// ===========================================================================

/// NETCONF error severity.
///
/// Matches `ErrorSeverity` in `NetconfABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ErrorSeverity {
    /// Error (tag 0).
    Error = 0,
    /// Warning (tag 1).
    Warning = 1,
}

impl ErrorSeverity {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Error),
            1 => Some(Self::Warning),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ErrorSeverity; 2] = [
        Self::Error, Self::Warning,
    ];
}

impl fmt::Display for ErrorSeverity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// NetconfState (tags 0-5)
// ===========================================================================

/// NETCONF session states.
///
/// Matches `NetconfState` in `NetconfABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum NetconfState {
    /// Idle (tag 0).
    Idle = 0,
    /// Connected (tag 1).
    Connected = 1,
    /// Locked (tag 2).
    Locked = 2,
    /// Editing (tag 3).
    Editing = 3,
    /// Closing (tag 4).
    Closing = 4,
    /// Terminated (tag 5).
    Terminated = 5,
}

impl NetconfState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Connected),
            2 => Some(Self::Locked),
            3 => Some(Self::Editing),
            4 => Some(Self::Closing),
            5 => Some(Self::Terminated),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [NetconfState; 6] = [
        Self::Idle, Self::Connected, Self::Locked, Self::Editing, Self::Closing, Self::Terminated,
    ];
}

impl fmt::Display for NetconfState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn netconf_operation_roundtrip() {
        for v in NetconfOperation::ALL {
            let tag = v.to_tag();
            let decoded = NetconfOperation::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(NetconfOperation::from_tag(12).is_none());
    }

    #[test]
    fn datastore_roundtrip() {
        for v in Datastore::ALL {
            let tag = v.to_tag();
            let decoded = Datastore::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Datastore::from_tag(3).is_none());
    }

    #[test]
    fn edit_operation_roundtrip() {
        for v in EditOperation::ALL {
            let tag = v.to_tag();
            let decoded = EditOperation::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(EditOperation::from_tag(5).is_none());
    }

    #[test]
    fn netconf_error_type_roundtrip() {
        for v in NetconfErrorType::ALL {
            let tag = v.to_tag();
            let decoded = NetconfErrorType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(NetconfErrorType::from_tag(4).is_none());
    }

    #[test]
    fn error_severity_roundtrip() {
        for v in ErrorSeverity::ALL {
            let tag = v.to_tag();
            let decoded = ErrorSeverity::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ErrorSeverity::from_tag(2).is_none());
    }

    #[test]
    fn netconf_state_roundtrip() {
        for v in NetconfState::ALL {
            let tag = v.to_tag();
            let decoded = NetconfState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(NetconfState::from_tag(6).is_none());
    }

    #[test]
    fn constants_match_idris() {
        assert_eq!(NETCONF_PORT, 830);
    }

}

File: protocols/proven-netconf/ffi/zig/src/netconf.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// netconf.zig -- Zig FFI implementation of proven-netconf.
//
// Implements the NETCONF (RFC 6241) session state machine with:
//   - 64-slot mutex-protected session pool
//   - Datastore lock tracking per session
//   - Edit-config operation queuing
//   - Candidate datastore validation and commit
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching NetconfABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching NetconfABI.Types.idr tag assignments)
// =========================================================================

/// NETCONF operations (ABI tags 0-11).
pub const Operation = enum(u8) {
    get = 0,
    get_config = 1,
    edit_config = 2,
    copy_config = 3,
    delete_config = 4,
    lock = 5,
    unlock = 6,
    close_session = 7,
    kill_session = 8,
    commit = 9,
    validate = 10,
    discard_changes = 11,
};

/// NETCONF datastores (ABI tags 0-2).
pub const Datastore = enum(u8) {
    running = 0,
    startup = 1,
    candidate = 2,
};

/// NETCONF edit operations (ABI tags 0-4).
pub const EditOperation = enum(u8) {
    merge = 0,
    replace = 1,
    create = 2,
    delete = 3,
    remove = 4,
};

/// NETCONF error types (ABI tags 0-3).
pub const ErrorType = enum(u8) {
    transport = 0,
    rpc = 1,
    protocol = 2,
    application = 3,
};

/// NETCONF error severity (ABI tags 0-1).
pub const ErrorSeverity = enum(u8) {
    err = 0,
    warning = 1,
};

/// NETCONF session lifecycle states (ABI tags 0-5).
pub const NetconfState = enum(u8) {
    idle = 0,
    connected = 1,
    locked = 2,
    editing = 3,
    closing = 4,
    terminated = 5,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent sessions.
const MAX_SESSIONS: usize = 64;

/// Maximum pending edit operations per session.
const MAX_EDITS: usize = 32;

/// Maximum hostname/xpath length in bytes.
const MAX_NAME_LEN: usize = 256;

/// A pending edit operation.
const EditEntry = struct {
    /// Target datastore.
    datastore: Datastore,
    /// Edit operation type.
    edit_op: EditOperation,
    /// XPath expression.
    xpath: [MAX_NAME_LEN]u8,
    xpath_len: u32,
    /// Whether this slot is active.
    active: bool,
};

/// A NETCONF session.
const Session = struct {
    /// Current lifecycle state.
    state: NetconfState,
    /// Remote host.
    host: [MAX_NAME_LEN]u8,
    host_len: u32,
    /// Remote port.
    port: u16,
    /// Which datastore is locked (if any). Only meaningful when state == .locked.
    locked_datastore: Datastore,
    /// Whether a datastore lock is held.
    has_lock: bool,
    /// Pending edit operations.
    edits: [MAX_EDITS]EditEntry,
    /// Number of pending edits.
    edit_count: u32,
    /// Total operations executed.
    op_count: u32,
    /// Whether this session slot is in use.
    active: bool,
};

/// Default (empty) edit entry.
const empty_edit: EditEntry = .{
    .datastore = .running,
    .edit_op = .merge,
    .xpath = [_]u8{0} ** MAX_NAME_LEN,
    .xpath_len = 0,
    .active = false,
};

/// Default (empty) session.
const empty_session: Session = .{
    .state = .idle,
    .host = [_]u8{0} ** MAX_NAME_LEN,
    .host_len = 0,
    .port = 830,
    .locked_datastore = .running,
    .has_lock = false,
    .edits = [_]EditEntry{empty_edit} ** MAX_EDITS,
    .edit_count = 0,
    .op_count = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

/// Returns the ABI version number.
pub export fn netconf_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new NETCONF session. Returns slot index (>=0) or -1 on failure.
/// The session starts in Connected state.
pub export fn netconf_create(
    host_ptr: [*]const u8,
    host_len: u32,
    port: u16,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (host_len == 0 or host_len > MAX_NAME_LEN) return -1;
    if (port == 0) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.host[0..host_len], host_ptr[0..host_len]);
            s.host_len = host_len;
            s.port = port;
            s.state = .connected;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a session, releasing its slot.
pub export fn netconf_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

/// Returns the current NetconfState tag for a session.
pub export fn netconf_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Lock a datastore. Connected -> Locked.
pub export fn netconf_lock(slot: c_int, datastore: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connected) return 1;
    if (datastore > 2) return 1;

    sessions[idx].locked_datastore = @enumFromInt(datastore);
    sessions[idx].has_lock = true;
    sessions[idx].state = .locked;
    sessions[idx].op_count += 1;
    return 0;
}

/// Unlock a datastore. Locked -> Connected.
pub export fn netconf_unlock(slot: c_int, datastore: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .locked) return 1;
    if (datastore > 2) return 1;
    if (@intFromEnum(sessions[idx].locked_datastore) != datastore) return 1;

    sessions[idx].has_lock = false;
    sessions[idx].state = .connected;
    sessions[idx].op_count += 1;
    return 0;
}

/// Get configuration from a datastore.
pub export fn netconf_get_config(slot: c_int, datastore: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connected and sessions[idx].state != .locked and
        sessions[idx].state != .editing) return 1;
    if (datastore > 2) return 1;

    sessions[idx].op_count += 1;
    return 0;
}

/// Edit configuration. Connected/Locked -> Editing.
pub export fn netconf_edit_config(
    slot: c_int,
    datastore: u8,
    edit_op: u8,
    xpath_ptr: [*]const u8,
    xpath_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connected and sessions[idx].state != .locked and
        sessions[idx].state != .editing) return 1;
    if (datastore > 2) return 1;
    if (edit_op > 4) return 1;
    if (xpath_len == 0 or xpath_len > MAX_NAME_LEN) return 1;
    if (sessions[idx].edit_count >= MAX_EDITS) return 1;

    // Queue the edit
    for (&sessions[idx].edits) |*e| {
        if (!e.active) {
            e.datastore = @enumFromInt(datastore);
            e.edit_op = @enumFromInt(edit_op);
            @memcpy(e.xpath[0..xpath_len], xpath_ptr[0..xpath_len]);
            e.xpath_len = xpath_len;
            e.active = true;
            sessions[idx].edit_count += 1;
            sessions[idx].state = .editing;
            sessions[idx].op_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Commit pending edits. Editing -> Connected (or Locked if lock held).
pub export fn netconf_commit(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .editing) return 1;

    // Clear all edits
    sessions[idx].edits = [_]EditEntry{empty_edit} ** MAX_EDITS;
    sessions[idx].edit_count = 0;
    sessions[idx].state = if (sessions[idx].has_lock) .locked else .connected;
    sessions[idx].op_count += 1;
    return 0;
}

/// Discard pending edits. Editing -> Connected (or Locked if lock held).
pub export fn netconf_discard(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .editing) return 1;

    sessions[idx].edits = [_]EditEntry{empty_edit} ** MAX_EDITS;
    sessions[idx].edit_count = 0;
    sessions[idx].state = if (sessions[idx].has_lock) .locked else .connected;
    sessions[idx].op_count += 1;
    return 0;
}

/// Validate a datastore.
pub export fn netconf_validate(slot: c_int, datastore: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connected and sessions[idx].state != .locked and
        sessions[idx].state != .editing) return 1;
    if (datastore > 2) return 1;

    sessions[idx].op_count += 1;
    return 0;
}

/// Close the session gracefully. Any non-Idle -> Closing.
pub export fn netconf_close_session(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .idle or state == .closing or state == .terminated) return 1;
    sessions[idx].state = .closing;
    sessions[idx].op_count += 1;
    return 0;
}

/// Kill another session (by session ID). Returns 0 on success.
pub export fn netconf_kill_session(slot: c_int, session_id: u32) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    _ = session_id;
    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connected and sessions[idx].state != .locked) return 1;
    sessions[idx].op_count += 1;
    return 0;
}

/// Complete cleanup. Closing/Terminated -> Idle.
pub export fn netconf_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .closing and sessions[idx].state != .terminated) return 1;

    sessions[idx].state = .idle;
    sessions[idx].has_lock = false;
    sessions[idx].edits = [_]EditEntry{empty_edit} ** MAX_EDITS;
    sessions[idx].edit_count = 0;
    sessions[idx].op_count = 0;
    return 0;
}

/// Check if a NETCONF state transition is valid (stateless).
pub export fn netconf_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Connected
    if (from == 1 and to == 2) return 1; // Connected -> Locked
    if (from == 2 and to == 1) return 1; // Locked -> Connected (unlock)
    if (from == 1 and to == 3) return 1; // Connected -> Editing
    if (from == 2 and to == 3) return 1; // Locked -> Editing
    if (from == 3 and to == 1) return 1; // Editing -> Connected (commit/discard)
    if (from == 3 and to == 2) return 1; // Editing -> Locked (commit/discard with lock)
    if (from == 1 and to == 4) return 1; // Connected -> Closing
    if (from == 2 and to == 4) return 1; // Locked -> Closing
    if (from == 3 and to == 4) return 1; // Editing -> Closing
    if (from == 4 and to == 0) return 1; // Closing -> Idle
    if (from == 5 and to == 0) return 1; // Terminated -> Idle
    return 0;
}

// --- pool size guard (audit S5: prevent oversized-global stack overflow) ---
comptime {
    if (@sizeOf(@TypeOf(sessions)) > 16 * 1024 * 1024)
        @compileError("pool 'sessions' exceeds the 16 MiB budget; heap-allocate or shrink (see audits/proof-panic-attack-2026-06-23.md)");
}
